packages/runner/injection/patches/utils/index.ts

```diff
@@ -0,0 +1,82 @@
+export const captureFullRequestUrl = (relativeOrAbsoluteUrlString: string, window: Window) => {
+  // need to pass the window here by reference to generate the correct absolute URL if needed. Spec Bridge does NOT contain sub domain
+  let url
+
+  try {
+    url = new URL(relativeOrAbsoluteUrlString).toString()
+  } catch (err1) {
+    try {
+      // likely a relative path, construct the full url
+      url = new URL(relativeOrAbsoluteUrlString, window.location.origin).toString()
+    } catch (err2) {
+      return undefined
+    }
+  }
+
+  return url
+}
+
+const CROSS_ORIGIN_PREFIX = 'cross:origin:'
+
+/**
+ * Sets up a promisified post message
+ * @param data the data to send
+ * @param event the name of the event to be promisified.
+ * @param timeout - in ms, if the promise does not complete during this timeout, fail the promise.
+ * @returns the data to send
+ */
+export const postMessagePromise = <T>({ event, data = {}, timeout }: {event: string, data: any, timeout: number}): Promise<T> => {
```

Comment:
this is very similar to what we're doing in the communicator, but just different enough to warrant it's own function.